contracts/cw-pre-propose-base/src/execute.rs

```diff
@@ -0,0 +1,223 @@
+
+        if !config.open_proposal_submission {
+            let dao = self.dao.load(deps.storage)?;
+            let voting_power: VotingPowerAtHeightResponse = deps.querier.query_wasm_smart(
+                dao.into_string(),
+                &CwCoreQuery::VotingPowerAtHeight {
+                    address: info.sender.to_string(),
+                    height: None,
+                },
+            )?;
+            if voting_power.power.is_zero() {
+                return Err(PreProposeError::NotMember {});
+            }
+        }
+
+        let deposit_messages = if let Some(ref deposit_info) = config.deposit_info {
+            deposit_info.check_native_deposit_paid(&info)?;
+            deposit_info.get_take_deposit_messages(&info.sender, &env.contract.address)?
+        } else {
+            vec![]
+        };
+
+        let proposal_module = self.proposal_module.load(deps.storage)?;
+        let propose_messsage = WasmMsg::Execute {
+            contract_addr: proposal_module.into_string(),
+            msg: to_binary(&msg)?,
+            funds: vec![],
+        };
+
+        Ok(Response::default()
+            .add_attribute("method", "execute_propose")
+            .add_attribute("sender", info.sender)
+            .add_attribute("deposit_info", to_binary(&config.deposit_info)?.to_string())
+            .add_messages(deposit_messages)
+            .add_message(propose_messsage))
+    }
+
+    pub fn execute_status_change_proposal_hook(
+        &self,
+        deps: Deps,
+        info: MessageInfo,
+        id: u64,
+        new_status: String,
+    ) -> Result<Response, PreProposeError> {
+        let proposal_module = self.proposal_module.load(deps.storage)?;
+        if info.sender != proposal_module {
+            return Err(PreProposeError::NotModule {});
+        }
+
+        let (deposit_info, proposer) = self.deposits.load(deps.storage, id)?;
+        let messages = if let Some(ref deposit_info) = deposit_info {
+            // If the proposal is completed, either return to the DAO
+            // or issue a refund.
+            let proposal_completed = new_status == "closed" || new_status == "executed";
+            // Refund can be issued if proposal if it is going to
+            // closed or executed.
+            let should_refund = (new_status == "closed"
+                && deposit_info.refund_policy == DepositRefundPolicy::Always)
+                || (new_status == "executed"
+                    && deposit_info.refund_policy != DepositRefundPolicy::Never);
+
+            if should_refund {
+                deposit_info.get_return_deposit_message(&proposer)?
+            } else if proposal_completed {
+                // If the proposer doesn't get the deposit, the DAO does.
+                let dao = self.dao.load(deps.storage)?;
+                deposit_info.get_return_deposit_message(&dao)?
+            } else {
+                vec![]
+            }
+        } else {
+            vec![]
+        };
```

Comment:
So, hear me out... what if we have two base contracts:
- `cw-pre-propose-base`: more general... no extra logic, used for custom pre-propose steps
- `cw-pre-propose-deposit-base`: uses `cw-pre-propose-base` for implementing deposit logic

There are other pre-propose flows that can be built like approvals or drafting.

Reply:
this is an interesting idea. I think you touch on a good point here that the current naming of this isn't great. IMO, we should rename what we have here to `cw-pre-propose-deposit-base`. I'm a little cautious about building a lower abstraction than that though yet as I feel that I'm not sure what it is.

For example: will all pre-propose modules need to interact with proposal status change hooks? What sort of instantiation information will they want? What sort of config options do they all share? I'm honestly pretty unsure. Maybe we wait on building that lower abstraction until we start writing more complex ones and can better identify shared logic?